File: backend/radar-service/src/main/java/com/radovan/spring/utils/ServiceUrlProvider.java
package com.radovan.spring.utils;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.radovan.spring.services.EurekaServiceDiscovery;

@Component
public class ServiceUrlProvider {

	private final Map<String, String> cachedServiceUrls = new ConcurrentHashMap<>();
	private EurekaServiceDiscovery eurekaServiceDiscovery;

	@Autowired
	private void initialize(EurekaServiceDiscovery eurekaServiceDiscovery) {
		this.eurekaServiceDiscovery = eurekaServiceDiscovery;
	}

	public String getServiceUrl(String serviceName) {
		return cachedServiceUrls.computeIfAbsent(serviceName, key -> {
			try {
				String serviceUrl = eurekaServiceDiscovery.getServiceUrl(serviceName);
				validateUrl(serviceUrl, serviceName);
				return serviceUrl;
			} catch (RuntimeException e) {
				System.err.println("Failed to retrieve service URL for: " + serviceName + " - " + e.getMessage());
				throw e;
			}
		});
	}

	public String getAuthServiceUrl() {
		return getServiceUrl("auth-service");
	}

	private void validateUrl(String url, String serviceName) {
		if (url == null || !url.startsWith("http")) {
			throw new IllegalArgumentException("Invalid URL for " + serviceName + ": " + url);
		}
	}

}

File: backend/api-gateway/src/main/scala/com/radovan/scalatra/utils/ServiceUrlProvider.scala
package com.radovan.scalatra.utils

import com.radovan.scalatra.services.EurekaServiceDiscovery
import jakarta.inject.Inject

import scala.collection.concurrent.TrieMap

class ServiceUrlProvider @Inject() (eurekaServiceDiscovery: EurekaServiceDiscovery) {

  private val cachedServiceUrls = TrieMap.empty[String, String]

  def getServiceUrl(serviceName: String): String = {
    cachedServiceUrls.getOrElseUpdate(serviceName, {
      val url = eurekaServiceDiscovery.getServiceUrl(serviceName)
      validateUrl(url, serviceName)
      url
    })
  }

  def getRadarServiceUrl: String = getServiceUrl("radar-service")
  def getRegistrationServiceUrl: String = getServiceUrl("registration-service")
  def getInfractionServiceUrl: String = getServiceUrl("infraction-service")
  def getApiGatewayUrl: String = getServiceUrl("api-gateway")
  def getAuthServiceUrl:String = getServiceUrl("auth-service")

  private def validateUrl(url: String, serviceName: String): Unit = {
    if (url == null || !url.startsWith("http")) {
      throw new IllegalArgumentException(s"Invalid URL for $serviceName: $url")
    }
  }
}